tools/chapel-py/chapel.cpp

```diff
@@ -0,0 +1,751 @@
+PLAIN_GETTER(OpCall, op, "s", return node->op());
+PLAIN_GETTER(Dot, field, "s", return node->field());
+PLAIN_GETTER(VisibilityClause, symbol, "O", return wrapAstNode(contextObject, node->symbol()));
+PLAIN_GETTER(Identifier, name, "s", return node->name());
+PLAIN_GETTER(SimpleBlockLike, block_style, "s", return blockStyleToString(node->blockStyle()));
+PLAIN_GETTER(Loop, block_style, "s", return blockStyleToString(node->blockStyle()));
+PLAIN_GETTER(EnumElement, init_expression, "O", return wrapAstNode(contextObject, node->initExpression()));
+
+
+static PyObject* AttributeObject_actuals(PyObject *self, PyObject *Py_UNUSED(ignored)) {
+  auto attributeNode = ((AttributeObject*) self)->parent.astNode->toAttribute();
+
+  auto argList = Py_BuildValue("(O)", (PyObject*) self);
+  auto astCallIterObjectPy = PyObject_CallObject((PyObject *) &AstCallIterType, argList);
+  Py_XDECREF(argList);
+
+  auto astCalliterObject = (AstCallIterObject*) astCallIterObjectPy;
+  astCalliterObject->current = 0;
+  astCalliterObject->num = attributeNode->numActuals();
+  astCalliterObject->container = attributeNode;
+  astCalliterObject->childGetter = [](const void* node, int child) {
+    return ((chpl::uast::Attribute*) node)->actual(child);
+  };
+  astCalliterObject->nameGetter = [](const void* node, int child) {
+    return ((chpl::uast::Attribute*) node)->actualName(child);
+  };
+
+  return astCallIterObjectPy;
+}
+
+static PyObject* FnCallObject_actuals(PyObject *self, PyObject *Py_UNUSED(ignored)) {
+  auto fnCallNode = ((FnCallObject*) self)->parent.astNode->toFnCall();
+
+  auto argList = Py_BuildValue("(O)", (PyObject*) self);
+  auto astCallIterObjectPy = PyObject_CallObject((PyObject *) &AstCallIterType, argList);
+  Py_XDECREF(argList);
+
+  auto astCalliterObject = (AstCallIterObject*) astCallIterObjectPy;
+  astCalliterObject->current = 0;
+  astCalliterObject->num = fnCallNode->numActuals();
+  astCalliterObject->container = fnCallNode;
+  astCalliterObject->childGetter = [](const void* node, int child) {
+    return ((chpl::uast::FnCall*) node)->actual(child);
+  };
+  astCalliterObject->nameGetter = [](const void* node, int child) {
+    return ((chpl::uast::FnCall*) node)->actualName(child);
+  };
+
+  return astCallIterObjectPy;
+}
+
+
+template <chpl::uast::asttags::AstTag tag>
+struct PerNodeInfo {
+  static constexpr PyMethodDef methods[] = {
+    {NULL, NULL, 0, NULL}  /* Sentinel */
+  };
+};
+
+#define METHOD_TABLE(TAG, ...)\
+  template <> \
+  struct PerNodeInfo<chpl::uast::asttags::TAG> {\
+    static constexpr PyMethodDef methods[] = {\
+      __VA_ARGS__ \
+      {NULL, NULL, 0, NULL}  /* Sentinel */ \
+    };\
+  }
+
+METHOD_TABLE(START_VarLikeDecl,
+  {"storage_kind", VarLikeDeclObject_storage_kind, METH_NOARGS, "Get the storage kind of this VarLikeDecl node"},
+);
+
+METHOD_TABLE(START_NamedDecl,
+  {"name", NamedDeclObject_name, METH_NOARGS, "Get the name of this NamedDecl node"},
+);
```

Comment:
As I read through this, I find myself wishing the `PLAIN_GETTER` and `METHOD_TABLE` where next to each other in the file

Reply:
As a second though, being able to have these defined in some kind of header would be fantastic. I think this one `.cpp` file is getting huge and unwieldy

Reply:
I'm still not sure how to structure this (which is why it's one file at the moment. I will give it some more thought in the next PR.